tests/e2e/utl_migrate_test.go

```diff
@@ -164,12 +186,104 @@ func TestEtctlutlMigrate(t *testing.T) {
 				}
 			}
 
-			t.Log("etcdutl migrate...")
+			t.Log("verify backend...")
 			be := backend.NewDefaultBackend(lg, filepath.Join(memberDataDir, "member/snap/db"))
-			defer be.Close()
 
 			ver := schema.ReadStorageVersion(be.ReadTx())
 			assert.Equal(t, tc.expectStorageVersion, ver)
+
+			for _, bk := range tc.expectNonFoundKeys {
+				_, vs := be.ReadTx().UnsafeRange(bk.bucket, bk.key, nil, 0)
+				assert.Zero(t, len(vs))
+			}
+
+			be.Close()
+
+			if tc.delWal {
+				// clear out v2store to avoid "etcdserver/membership: cluster cannot be downgraded (current version: X is lower than determined cluster version: X)"
```

Comment:
Don't think this is correct. Reason `etcdserver/membership: cluster cannot be downgraded` is written is because etcd server will still replay wal entries from since last snapshot. Proper solution would be to ensure that there is no `SetClusterVersion` wal entries since last snapshot.

Reply:
quick update: working on adding snapshot save/restore for the case when `/0/version` entry is in wal log. I think that's the best way to go about it, since that's how users will need to handle this.

Reply:
> quick update: working on adding snapshot save/restore for the case when /0/version entry is in wal log

Don't understand. You mean to use snapshot save/restore to workaround the issue. I don't think it's correct, the only reason it works is fact that it throws out the WAL completely. Don't think this is correct in 3 node clusters as there is no guarantee that snapshot will save the same data on different nodes.

Proper solution would be to modify/remove the SetClusterVersion WAL entry. Generally it can be risky, as this entry protects etcd from writing incompatible WAL entries that would be interpreted wrongly. There were some new fields introduced into etcd v3.5 WAL. List is available in https://github.com/etcd-io/etcd/blob/main/scripts/etcd_version_annotations.txt. 

For v3.6 downgrades I implemented a mechanism to detect WAL version based on proto annotations https://github.com/etcd-io/etcd/blob/main/server/storage/wal/version.go. However maybe we could check what proto versions were between v3.4 and v3.4, maybe the differences are not critical so we could yolo it.

Reply:
@serathius you are correct, one has to use same snapshot. I was thinking of something like [this](https://github.com/etcd-io/etcd/issues/15878#issuecomment-1571620392) for 3 node cluster. There will be cluster downtime if using this approach. There is probably a chance of data loss for entries that are in wal but not applied to backend (maybe there is a way to mitigate this).

Updating WAL is probably ideal. [Here](https://www.diffnow.com/report/2fnge) is the diff of raft_internal. From the first glance all new requests can be removed during migrate.`ClusterVersionSetRequest`, `ClusterMemberAttrSetRequest`, `DowngradeInfoSetRequest` and `AuthStatusRequest`. I still need to check every existing request for possible new non-optional fields.

I'll give it a shot and try to add wal update to `migrate`.

Reply:
Please note that each member could be on different stage of applying WAL and storing db. We need to make sure that applying those entries works the same on both v3.5/v3.4.